File: 211-DesignAddAndSearchWordsDataStructure/211-DesignAddAndSearchWordsDataStructure.js
// Last updated: 4/29/2026, 10:32:55 AM
/**
 * Initialize your data structure here.
 */
var WordDictionary = function() {
    this.trie = {};
};

/** 
 * @param {string} word
 * @return {void}
 */
WordDictionary.prototype.addWord = function(word) {
    let node = this.trie;
    for(const ch of word) {
        if(!node[ch]) {
            node[ch] = {};
        } 
        node = node[ch];
    }
    node.word = true;
    
    // console.log(this.trie);
};

/** 
 * @param {string} word
 * @return {boolean}
 */
WordDictionary.prototype.search = function(word) {
    let node = this.trie;
    
    function find(curr, remain, node) {
        if(remain.length === 0) {
            if(node.word) {
                return true;
            } else {
                return false;
            }
        } 
        const ch = remain[0];
        if(ch === '.') {
            const keys = Object.keys(node);
            for(const key of keys) {
                if(key === 'word') {
                    continue;
                }
                const newCurr = curr.concat(key);
                const newRemain = remain.slice(1);
                if(find(newCurr, newRemain, node[key])) {
                    return true;
                }
            }    
        } else {
            if(node[ch]) {
                const newCurr = curr.concat(ch);
                const newRemain = remain.slice(1);
                return find(newCurr, newRemain, node[ch]);    
            } else {
                return false;
            }
        }
        return false;
    }
    
    return find('', word, node);
};

/** 
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = new WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */

File: 211-DesignAddAndSearchWordsDataStructure/211-DesignAddAndSearchWordsDataStructure.py
# Last updated: 4/29/2026, 10:32:50 AM
class WordDictionary:

    def __init__(self):
        self.root = {}

    def addWord(self, word: str) -> None:
        node = self.root
        for ch in word:
            if node.get(ch) == None:
                node[ch] = {}
            node = node[ch]
        node['completed'] = word
        # print(self.root)

    def search(self, word: str) -> bool:
        node = self.root
        queue = [ node ]
        idx = 0
        while len(queue) != 0 and idx < len(word):
            next_queue = []
            ch = word[idx]
            # print(ch)
            for node in queue:
                if ch == '.':
                    for next_ch in node:
                        if next_ch != 'completed':
                            next_queue.append(node[next_ch])
                else:
                    if node.get(ch) != None:
                        next_queue.append(node[ch])
            queue = next_queue
            idx += 1
        
        if idx == len(word):
            for node in queue:
                if node.get('completed') != None:
                    # print("return True")
                    return True
        else:
            # print("return False")
            return False
        
                
            
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)